src/dbnode/storage/repair.go

```diff
@@ -426,12 +447,79 @@ func (r *dbRepairer) Repair() error {
 	if err != nil {
 		return err
 	}
+
 	for _, n := range namespaces {
-		iter := r.namespaceRepairTimeRanges(n).Iter()
-		for iter.Next() {
-			multiErr = multiErr.Add(r.repairNamespaceWithTimeRange(n, iter.Value()))
+		repairRange := r.namespaceRepairTimeRange(n)
+		blockSize := n.Options().RetentionOptions().BlockSize()
+
+		// Iterating backwards will be exclusive on the start, but we want to be inclusive on the
+		// start so subtract a blocksize.
+		repairRange.Start = repairRange.Start.Add(-blockSize)
+
+		numUnrepairedBlocks := 0
+		repairRange.IterateBackwards(blockSize, func(blockStart time.Time) bool {
+			repairState, ok := r.repairStatesByNs.repairStates(n.ID(), blockStart)
+			if !ok || repairState.Status != repairSuccess {
+				numUnrepairedBlocks++
+			}
+
+			return true
+		})
+
+		r.scope.Tagged(map[string]string{
+			"namespace": n.ID().String(),
+		}).Gauge("num-unrepaired-blocks").Update(float64(numUnrepairedBlocks))
+
+		if numUnrepairedBlocks > 0 {
+			repairRange.IterateBackwards(blockSize, func(blockStart time.Time) bool {
+				repairState, ok := r.repairStatesByNs.repairStates(n.ID(), blockStart)
+				if !ok || repairState.Status != repairSuccess {
+					if err := r.repairNamespaceBlockstart(n, blockStart); err != nil {
+						multiErr = multiErr.Add(err)
+					}
+					// Only repair one block per namespace per iteration.
+					return false
+				}
+				return true
+			})
+
+			continue
+		}
+
+		var (
+			leastRecentlyRepairedBlockStart               time.Time
+			leastRecentlyRepairedBlockStartLastRepairTime time.Time
+			maxSecondsSinceLastBlockRepair                = r.scope.Tagged(map[string]string{
+				"namespace": n.ID().String(),
+			}).Gauge("max-seconds-since-last-block-repair")
+		)
+		repairRange.IterateBackwards(blockSize, func(blockStart time.Time) bool {
+			repairState, ok := r.repairStatesByNs.repairStates(n.ID(), blockStart)
+			if !ok {
+				// Should never happen.
+				instrument.EmitAndLogInvariantViolation(r.opts.InstrumentOptions(), func(l *zap.Logger) {
+					l.With(
+						zap.Time("blockStart", blockStart),
+						zap.String("namespace", n.ID().String()),
+					).Error("missing repair state in all-blocks-are-repaired branch")
+				})
+				return true
+			}
+
+			if leastRecentlyRepairedBlockStart.IsZero() || repairState.LastAttempt.Before(leastRecentlyRepairedBlockStartLastRepairTime) {
```

Comment:
Seems like this state can also be calculated in the same loop above, but only emit the gauge/repair if `numUnrepairedBlocks == 0`